discovery/openstack/instance.go

```diff
@@ -0,0 +1,225 @@
+	openstackLabelPublicIP       = openstackLabelPrefix + "public_ip"
+	openstackLabelPrivateIP      = openstackLabelPrefix + "private_ip"
+	openstackLabelTagPrefix      = openstackLabelPrefix + "tag_"
+)
+
+// InstanceDiscovery discovers OpenStack instances.
+type InstanceDiscovery struct {
+	authOpts *gophercloud.AuthOptions
+	region   string
+	interval time.Duration
+	logger   log.Logger
+	port     int
+}
+
+// NewInstanceDiscovery returns a new instance discovery.
+func NewInstanceDiscovery(opts *gophercloud.AuthOptions,
+	interval time.Duration, port int, region string, l log.Logger) *InstanceDiscovery {
+	return &InstanceDiscovery{authOpts: opts,
+		region: region, interval: interval, port: port, logger: l}
+}
+
+// Run implements the TargetProvider interface.
+func (i *InstanceDiscovery) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	// Get an initial set right away.
+	tg, err := i.refresh()
+	if err != nil {
+		i.logger.Error(err)
+	} else {
+		select {
+		case ch <- []*config.TargetGroup{tg}:
+		case <-ctx.Done():
+			return
+		}
+	}
+
+	ticker := time.NewTicker(i.interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			tg, err := i.refresh()
+			if err != nil {
+				i.logger.Error(err)
+				continue
+			}
+
+			select {
+			case ch <- []*config.TargetGroup{tg}:
+			case <-ctx.Done():
+				return
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (i *InstanceDiscovery) refresh() (*config.TargetGroup, error) {
+	var err error
+	t0 := time.Now()
+	defer func() {
+		refreshDuration.Observe(time.Since(t0).Seconds())
+		if err != nil {
+			refreshFailuresCount.Inc()
+		}
+	}()
+
+	provider, err := openstack.AuthenticatedClient(*i.authOpts)
+	if err != nil {
+		return nil, fmt.Errorf("could not create OpenStack session: %s", err)
+	}
+	client, err := openstack.NewComputeV2(provider, gophercloud.EndpointOpts{
+		Region: i.region,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("could not create OpenStack compute session: %s", err)
+	}
+
+	opts := servers.ListOpts{}
+	pager := servers.List(client, opts)
+
+	tg := &config.TargetGroup{
+		Source: fmt.Sprintf("OS_" + i.region),
+	}
+
+	pagerFIP := floatingips.List(client)
+	floatingIPList := make(map[string][]string)
+
+	err = pagerFIP.EachPage(func(page pagination.Page) (bool, error) {
+		result, err := floatingips.ExtractFloatingIPs(page)
+		if err != nil {
+			i.logger.Warn(err)
+		}
+		for _, ip := range result {
+			// Skip not associated ips
+			if ip.InstanceID != "" {
+				floatingIPList[ip.InstanceID] = append(floatingIPList[ip.InstanceID], ip.IP)
+			}
+		}
+		return true, nil
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("could not describe floating IPs: %s", err)
+	}
+
+	err = pager.EachPage(func(page pagination.Page) (bool, error) {
+		serverList, err := servers.ExtractServers(page)
+		if err != nil {
+			return false, fmt.Errorf("could not extract servers: %s", err)
```

Comment:
So why are the terms server and instance used interchangeably? Should this discovery be called ServerDiscovery? Or is instance a synonym for server in OpenStack? In either way, we should settle on one term.

Reply:
Updated.  settle on one term **instance**